### ISSUE TITLE
Reach of the reinit() operator

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#2038]
**Reported by choeger on 5 Aug 2016 10:46 UTC**
In a model with two variables:

```mo
model A
function f input Real x; output Real y; /*arbitrarily complex with f(1.0) = 0 */ end f;
Real x,y;
equation
der(x) = 1.;
y = f(x);

when time > 0.5 then 
  reinit(y, 0);
end when;

when x > 1. then
  reinit(x, 0);
end when;

end A;
```

The specification tells us (section 8.3.6) that `reinit(y,0)`  makes `y` unknown and introduces the equation `y = 0`.

Does that mean that during event iteration:
 * `0=f(x)` holds, i.e. `x` has to be recalculated by finding a root of `f` in order to check for the second when-clause
 * The second when-clause will never be triggered by the first when-clause

The second option is much simpler to implement and I do not know if there is any use for the first option, but if the first option is required, the spec should explicitly state how to make a variable "unknown" and how to solve the resulting system of equations.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/2038

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 7 Sep 2016 09:24 UTC**
Replying to [ticket:2038 choeger]:
> The specification tells us (section 8.3.6) that `reinit(y,0)`  makes `y` unknown and introduces the equation `y = 0`.

This is no longer the case in 3.3r1 and in the model above both x and y cannot be states (since `y=f(x)` ).

The new statement is:
 The reinit operator does not break the single assignment rule, because reinit(x,expr) in equations evaluates expr to a value (value), then at the end of the current event iteration step it assigns this value to x (this copying from values to reinitialized state(s) is done after all other evaluations of the model and before copying x to pre(x)).

So basically this is a duplicate of the issue in #1372

**Comment by modelica-trac-importer** [comment]
**Comment by henrikt on 12 Sep 2016 08:56 UTC**
The new formulation is a rather awkward way of giving the details of the semantics.  Saying _because_ sounds like the sentence is just about to rephrase what has already been said.  It would be better to give the semantics right at the start:
  The operator reinitializes x with expr at an event instant, by replacing the current value of `x` with a new value at the end of the current event iteration.  In the next event iteration, `pre(x)` refers to the new value.

Once this is said (carefully avoiding the word _assign_ to avoid confusion), a note on the single assignment rule may clarify that the replacement action of `reinit` does not count as an assignment in the single assignment rule.
